consensus/src/twins_test.rs

```diff
@@ -200,3 +235,193 @@ fn basic_start_test() {
+        .config
+        .validator_network
+        .as_ref()
+        .unwrap()
+        .peer_id();
+    // n2 will be the leader (because it has the lowest hash)
+    let n2 = nodes[2]
+        .config
+        .validator_network
+        .as_ref()
+        .unwrap()
+        .peer_id();
+    let n3 = nodes[3]
+        .config
+        .validator_network
+        .as_ref()
+        .unwrap()
+        .peer_id();
+
+    let n0_twin_id = *playground.get_twin_ids(n0).get(0).unwrap();
+    let n1_twin_id = *playground.get_twin_ids(n1).get(0).unwrap();
+    let n2_twin_id = *playground.get_twin_ids(n2).get(0).unwrap();
+    let n3_twin_id = *playground.get_twin_ids(n3).get(0).unwrap();
+
+    playground.split_network(vec![n1_twin_id], vec![n0_twin_id, n2_twin_id, n3_twin_id]);
+
+    timed_block_on(&mut runtime, async {
+        //timed_block_on(&mut runtime, async {
+        let _proposal = playground
+            .wait_for_messages(1, NetworkPlayground::proposals_only)
+            .await;
+
+        // Pull enough votes to get a few commits
+        // The proposer's votes are implicit and do not go in the queue.
+        let _votes = playground
+            .wait_for_messages(30, NetworkPlayground::votes_only)
+            .await;
+
+        // Check that the commit log for n0 is not empty
+        let mut commit_seen = false;
+        nodes[0].commit_cb_receiver.close();
+        if let Ok(Some(_node_commit)) = nodes[0].commit_cb_receiver.try_next() {
+            commit_seen = true;
+        }
+        assert!(commit_seen);
+
+        // Check that the commit log for n1 is empty
+        commit_seen = false;
+        nodes[1].commit_cb_receiver.close();
+        if let Ok(Some(_node_commit)) = nodes[1].commit_cb_receiver.try_next() {
+            commit_seen = true;
+        }
+        assert!(!commit_seen);
+    });
+}
+
+#[test]
+/// This test checks that the vote of a node and its twin
+/// should be counted as duplicate vote (because they have
+/// the same public keys)
+///
+/// Setup:
+///
+/// 4 honest nodes (n0, n1, n2, n3), and 1 twin (twin0)
+/// Create 2 partitions, p1=[n1, n3], p2=[n0, twin0, n2]
+///
+/// Test:
+///
+/// Extract enough votes to potentially form commits. Check
+/// that no node commits any block. This is because we need
+/// 3 nodes to form a quorum and no partition has enough votes
+/// (note there are 3 nodes in p2, but one of them is a twin,
+/// and its vote will be counted as duplicate of n0).
+///
+/// Run the test:
+/// cargo xtest -p consensus twins_vote_dedup_test -- --nocapture
+#[cfg(test)]
+fn twins_vote_dedup_test() {
+    let mut runtime = consensus_runtime();
+    let mut playground = NetworkPlayground::new(runtime.handle().clone());
+    let num_nodes = 4;
+    let num_twins = 1;
+    let mut nodes = SMRNode::start_num_nodes_with_twins(
+        num_nodes,
+        num_twins,
+        &mut playground,
+        RotatingProposer,
+    );
+
+    // 4 honest nodes
+    let n0 = nodes[0]
```

Comment:
Given this is boilerplate, would it make sense to have a helper function that returns the peer ids in a Vec or start_num_nodes_with_twins could return (nodes, peer_ids)?

Reply:
Good idea! Went for the second option i.e. start_num_nodes_with_twins now returns (nodes, node_authors).